libs/auth/src/apdu.ts

```diff
@@ -262,24 +267,54 @@ export function constructTlv(
 export function parseTlv(
   tagAsByteOrBuffer: Byte | Buffer,
   tlv: Buffer
-): [Buffer, Buffer, Buffer] {
-  const tagLength = Buffer.isBuffer(tagAsByteOrBuffer)
-    ? tagAsByteOrBuffer.length
-    : 1;
+): [tag: Buffer, length: Buffer, value: Buffer] {
+  const expectedTagBytes = Buffer.isBuffer(tagAsByteOrBuffer)
+    ? tagAsByteOrBuffer
+    : Buffer.of(tagAsByteOrBuffer);
+  const expectedTagLength = expectedTagBytes.length;
+  const tagBytes = tlv.subarray(0, expectedTagLength);
+  assert(
+    tagBytes.equals(expectedTagBytes),
+    `TLV tag (${inspect(tagBytes)}) does not match expected tag (${inspect(
+      expectedTagBytes
+    )})`
+  );
 
-  let tlvLengthLength = 1; // Derpiest variable name, I know
-  const tlvLengthFirstByte = tlv.at(tagLength);
+  let tlvLengthLength: number;
+  let tlvLength: number;
+  const tlvLengthFirstByte = assertDefined(tlv.at(expectedTagLength));
+  const tlvLengthSecondByte = tlv.at(expectedTagLength + 1);
+  const tlvLengthThirdByte = tlv.at(expectedTagLength + 2);
   if (tlvLengthFirstByte === 0x81) {
     tlvLengthLength = 2;
+    tlvLength = assertDefined(tlvLengthSecondByte);
   } else if (tlvLengthFirstByte === 0x82) {
     tlvLengthLength = 3;
+    tlvLength =
+      // eslint-disable-next-line no-bitwise
+      (assertDefined(tlvLengthSecondByte) << 8) +
+      assertDefined(tlvLengthThirdByte);
+  } else if (tlvLengthFirstByte <= 0x80) {
+    tlvLengthLength = 1;
+    tlvLength = assertDefined(tlvLengthFirstByte);
+  } else {
+    throw new Error(
+      `TLV length is invalid: received 0x${tlvLengthFirstByte.toString(
+        16
+      )}, but expected a value <= 0x80`
+    );
```

Comment:
Since `0x81` and `0x82` are valid first length bytes as well, a message like this might be preferable:

> TLV length is invalid: Received 0xXX but expected a value <= 0x82 for the first length byte